using float/int sumtype cast results in C error

writing a bug report due to an issue posted in the discord. [message link](https://discord.com/channels/592103645835821068/592114487759470596/975857251334115408)

**What did you do?**
a.v
```v
fn cast(a int | f64) f64 {
	return f64(a)
}

println(cast(10))
```
`v run a.v`

**What did you expect to see?**
`10.` in the console
 
**What did you see instead?**
```
/tmp/v_1000/a.18217470816788589904.tmp.c:11611: error: invalid aggregate type for register load
builder error:
==================
```
which is the C representation of the function:

```c
f64 main__cast(main___v_anon_sum_type_f64_int a) {
	f64 _t1 = ((f64)(a));
	return _t1;
}
```

VS Code squiggles this cast, so i suspect this to be a cgen error. using a non-anonymous sumtype results in the same error and C code.

V version: `V 0.2.4 bc397bb`
OS: Arch Linux 5.16.16-arch-1
Compiler used: tcc

--- a/vlib/json/json_omitempty_types_test.v
+++ b/vlib/json/json_omitempty_types_test.v
@@ -45,83 +45,83 @@ fn main() {
 }
 
 fn test_struct() {
-	test2 := OmitEmptyStruct{
+	test := OmitEmptyStruct{
 		bug: Struct{}
 	}
-	encoded2 := json.encode(test2)
-	dump(encoded2)
-	test := OmitEmptyStruct{
+	encoded := json.encode(test)
+	dump(encoded)
+	test2 := OmitEmptyStruct{
 		bug: Struct{
 			name: 'mybug'
 		}
 	}
-	encoded := json.encode(test)
-	dump(encoded)
+	encoded2 := json.encode(test2)
+	dump(encoded2)
 }
 
 fn test_fnum_struct() {
 	test := OmitEmptyFNumStruct{
 		bug: FNumStruct{}
 	}
-	encode := json.encode(test)
-	dump(encode)
+	encoded := json.encode(test)
+	dump(encoded)
 	test2 := OmitEmptyFNumStruct{
 		bug: FNumStruct{
 			f_num: 1.5
 		}
 	}
-	encoded := json.encode(test)
-	dump(encoded)
+	encoded2 := json.encode(test)
+	dump(encoded2)
 }
 
 fn test_alias() {
-	test3 := OmitEmptyAlias{
+	test := OmitEmptyAlias{
 		bug: ''
 	}
-	encoded3 := json.encode(test3)
-	dump(encoded3)
-	test4 := OmitEmptyAlias{
+	encoded := json.encode(test)
+	dump(encoded)
+	test2 := OmitEmptyAlias{
 		bug: 'foo'
 	}
-	encoded4 := json.encode(test4)
-	dump(encoded4)
+	encoded2 := json.encode(test2)
+	dump(encoded2)
 }
 
 fn test_map() {
-	test3 := OmitEmptyMap{
+	test := OmitEmptyMap{
 		bug: {}
 	}
-	encoded3 := json.encode(test3)
-	dump(encoded3)
-	test4 := OmitEmptyMap{
+	encoded := json.encode(test)
+	dump(encoded)
+	test2 := OmitEmptyMap{
 		bug: {
 			'foo': 'bar'
 		}
 	}
-	encoded4 := json.encode(test4)
-	dump(encoded4)
+	encoded2 := json.encode(test2)
+	dump(encoded2)
 }
 
 fn test_sumtype() {
-	test3 := OmitEmptySumType{}
-	encoded3 := json.encode(test3)
-	dump(encoded3)
-	test4 := OmitEmptySumType{
+	test := OmitEmptySumType{}
+	encoded := json.encode(test)
+	dump(encoded)
+	test2 := OmitEmptySumType{
 		bug: 1
 	}
-	encoded4 := json.encode(test4)
-	dump(encoded4)
+	encoded2 := json.encode(test2)
+	dump(encoded2)
 }
 
 fn test_array() {
-	test3 := OmitEmptyArray{
+	test := OmitEmptyArray{
 		bug: []
 	}
-	encoded3 := json.encode(test3)
-	dump(encoded3)
-	test4 := OmitEmptyArray{
+	encoded := json.encode(test)
+	dump(encoded)
+	test2 := OmitEmptyArray{
 		bug: ['1']
 	}
-	encoded4 := json.encode(test4)
-	dump(encoded4)
+	encoded2 := json.encode(test2)
+	dump(encoded2)
 }

--- a/vlib/json/json_omitempty_types_test.v
+++ b/vlib/json/json_omitempty_types_test.v
@@ -62,7 +62,7 @@ fn test_fnum_struct() {
 			f_num: 1.5
 		}
 	}
-	encoded2 := json.encode(test)
+	encoded2 := json.encode(test2)
 	dump(encoded2)
 }
 

--- a/vlib/json/json_omitempty_types_test.v
+++ b/vlib/json/json_omitempty_types_test.v
@@ -36,14 +36,6 @@ struct OmitEmptyFNumStruct {
 	bug FNumStruct [omitempty]
 }
 
-fn main() {
-	test_struct()
-	test_alias()
-	test_map()
-	test_sumtype()
-	test_array()
-}
-
 fn test_struct() {
 	test := OmitEmptyStruct{
 		bug: Struct{}

--- a/vlib/json/json_omitempty_types_test.v
+++ b/vlib/json/json_omitempty_types_test.v
@@ -28,6 +28,14 @@ struct OmitEmptySumType {
 	bug MySum [omitempty]
 }
 
+struct FNumStruct {
+	f_num f64
+}
+
+struct OmitEmptyFNumStruct {
+	bug FNumStruct [omitempty]
+}
+
 fn main() {
 	test_struct()
 	test_alias()
@@ -45,6 +53,21 @@ fn test_struct() {
 	test := OmitEmptyStruct{
 		bug: Struct{
 			name: 'mybug'
+		}
+	}
+	encoded := json.encode(test)
+	dump(encoded)
+}
+
+fn test_fnum_struct() {
+	test := OmitEmptyFNumStruct{
+		bug: FNumStruct{}
+	}
+	encode := json.encode(test)
+	dump(encode)
+	test2 := OmitEmptyFNumStruct{
+		bug: FNumStruct{
+			f_num: 1.5
 		}
 	}
 	encoded := json.encode(test)

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -6116,7 +6116,7 @@ fn (mut g Gen) type_default(typ_ ast.Type) string {
 				for field in info.fields {
 					field_sym := g.table.sym(field.typ)
 					if field.has_default_expr
-						|| field_sym.kind in [.array, .map, .string, .bool, .alias, .i8, .i16, .int, .i64, .u8, .u16, .u32, .u64, .char, .voidptr, .byteptr, .charptr, .struct_, .chan] {
+						|| field_sym.kind in [.array, .map, .string, .bool, .alias, .i8, .i16, .int, .i64, .u8, .u16, .u32, .u64, .f32, .f64, .char, .voidptr, .byteptr, .charptr, .struct_, .chan] {
 						field_name := c_name(field.name)
 						if field.has_default_expr {
 							mut expr_str := ''

